From 5b500882bb51cff0f38a51735bf008923cde2219 Mon Sep 17 00:00:00 2001
From: Julien Portalier <julien@portalier.com>
Date: Mon, 1 Jul 2024 17:31:15 +0200
Subject: [PATCH] Fix: don't hardcode alpn protocol byte size (openssl)
 (#14769)

For some reason OpenSSL used to negotiate the protocol by itself,
without invoking the select callback, or maybe it didn't respect the
total bytesize when processing the alpn string.

That changed in the 3.0.14 and other bugfix releases of OpenSSL, which
exposed the bug.
---
 src/openssl/ssl/context.cr | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/openssl/ssl/context.cr b/src/openssl/ssl/context.cr
index c7d5b5a0de2a..38e0054cba17 100644
--- a/src/openssl/ssl/context.cr
+++ b/src/openssl/ssl/context.cr
@@ -178,7 +178,7 @@ abstract class OpenSSL::SSL::Context
       {% if LibSSL.has_method?(:ssl_ctx_set_alpn_select_cb) %}
         alpn_cb = ->(ssl : LibSSL::SSL, o : LibC::Char**, olen : LibC::Char*, i : LibC::Char*, ilen : LibC::Int, data : Void*) {
           proto = Box(Bytes).unbox(data)
-          ret = LibSSL.ssl_select_next_proto(o, olen, proto, 2, i, ilen)
+          ret = LibSSL.ssl_select_next_proto(o, olen, proto, proto.size, i, ilen)
           if ret != LibSSL::OPENSSL_NPN_NEGOTIATED
             LibSSL::SSL_TLSEXT_ERR_NOACK
           else
